### ISSUE TITLE
Support lambda literals for non-delegate types?

### ISSUE THREAD

**Issue by eiriktsarpalis** [dotnet/roslyn#11929]
We at the F# community have been investigating ways in which [F# APIs could be made easier to consume from C# code](https://github.com/Microsoft/visualfsharp/issues/1254). One of the primary pain points is that it currently is difficult to define literals for the [F# function type](https://msdn.microsoft.com/en-us/visualfsharpdocs/conceptual/core.fsharpfunc%5B't,'u%5D-class-%5Bfsharp%5D). For those not familiar with F# functions, they are defined as classes inheriting the abstract type [`FSharpFunc<T,U>`](https://msdn.microsoft.com/en-us/visualfsharpdocs/conceptual/core.fsharpfunc%5B't,'u%5D-class-%5Bfsharp%5D) defined in the `FSharp.Core` library. This has been done to facilitate partial application, a feature common in functional languages.

The F# core library provides [a set of implicit converters](https://msdn.microsoft.com/en-us/visualfsharpdocs/conceptual/fsharpfunc.op_implicit['t,'u]-method-[fsharp]) between the `System.Converter<T,U>` delegate and `FSharpFunc<T,U>`. One would expect that it should make it easy to declare F# function literals in C# like so:
```csharp
FSharpFunc<int,int> func = x => x + 1;
```
However the code above fails with the message:
```csharp
error CS1660: Cannot convert lambda expression to type 'FSharpFunc<int, int>' because it is not a delegate type
```
The error can be overcome by adding an explicit cast to the lamda
```csharp
FSharpFunc<int,int> func = (Converter<int,int>)(x => x + 1);
```
which seems bizarre.

Would you consider enabing support for lambda literals in types that are not delegates?

**Comment by HaloFour** [comment]
How is this type generally consumed from C#?  I assume that C# would be passing said function to a method of some exposed type?  What kind of convention could C# recognize as a "lambda-like" as to not take any direct dependencies on any specific F#-types or conventions, and could F# adapt to a newly specified convention without breaking existing source?

**Comment by eiriktsarpalis** [comment]
Any F# API that takes lambdas as parameters (think LINQ) is impractical to consume from a C# point of view. 

One possibility would be to have the compiler infer the lambda literal type based on available implicit converters from delegate types to the assigned type at hand. This should work assuming no ambiguities arise.

**Comment by HaloFour** [comment]
Seems reasonable.  

```cs
FSharpFunc<int,int> func = x => x + 1;
// translates into
FSharpFunc<int, int> func = FSharpFunc<int, int>.op_Implicit((Converter<int, int>)(x => x + 1));
```

**Comment by alrz** [comment]
Related to #4801

**Comment by HaloFour** [comment]
@alrz 

That's true, such a modification to how lambdas are handled could introduce subtle breaking changes when it comes to overloading.  I'd think that implicit operators would only be considered after any overloads with delegates have been ruled out.

**Comment by gulshan** [comment]
I think, there should be a static method `fun` converting C# lambda to F# functions like-
``` csharp
using Microsoft.FSharp.Core.FSharpFunc;

var inc = fun(x => x + 1);
```

Type inferece will be an issue in this case though as @DavidArno mentioned here regarding implementation of similar feature in his SuccincT library-   
https://github.com/DavidArno/SuccincT/wiki/TypedLambdas

In that case, the syntax will be either of these two-
``` csharp
var inc = fun<int>(x => x + 1);
/*or*/
var inc = fun((int x) => x + 1);
```
IMHO, C# team should be focusing more on type inference for cases like this.
